### PR TITLE
Tolerate unsolicited HTTP responses.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

Right now NIO clients will crash in the (admittedly unlikely) event that
a HTTP server sends an unsolicited extra response on a connection. This
seems pretty unnecessary, and fairly graceless.

Modifications:

- Replace a forcible pop with an optional one.
- Added NIOHTTPDecoderError type.

Result:

Removed a crash that we definitely don't want.

### CODE REVIEW

File: Tests/NIOHTTP1Tests/HTTPDecoderTest.swift [hunk 1]
```diff
@@ -745,4 +745,17 @@ class HTTPDecoderTest: XCTestCase {
         XCTAssertNoThrow(XCTAssert(try channel.finish().isClean))
     }
 
+    func testAppropriateErrorWhenReceivingUnsolicitedResponse() throws {
```
Comment (weissi) [inline Tests/NIOHTTP1Tests/HTTPDecoderTest.swift]: could we add a test case that verifies that we also cannot silently 'recover' from having a bad server? Ie:

- `-> HTTP/1.1 200 OK\r\nServer: a-bad-server/1.0.0\r\n\r\n`
- `<- GET / HTTP/1.1`
- `-> HTTP/1.1 200 OK\r\nServer: a-bad-server/1.0.0\r\n\r\n`

isn't accepted

Comment (Lukasa) [reply Tests/NIOHTTP1Tests/HTTPDecoderTest.swift]: I think right now it absolutely will be. Do we need to defend against that behaviour? This seems like the kind of thing we normally leave to having a pipeline error handler.

Comment (weissi) [reply Tests/NIOHTTP1Tests/HTTPDecoderTest.swift]: @Lukasa I think we should protect. Not that we should close the connection but I believe we should consistently error. A server that starts responding before a request isn't quite right and I think we should continue sending errors. But happy to discuss that.

Comment (Lukasa) [reply Tests/NIOHTTP1Tests/HTTPDecoderTest.swift]: Ok, test added to verify.

**Pull request merged**